irr/include/irrString.h

```diff
@@ -173,13 +173,16 @@ class string
 			return *this;
 		}
 
-		// no longer allowed!
-		_IRR_DEBUG_BREAK_IF((void *)c == (void *)c_str());
-
-		u32 len = calclen(c);
-		str.resize(len);
-		for (u32 l = 0; l < len; ++l)
-			str[l] = (T)c[l];
+        if ((void*)c == (void*)c_str())
```

Comment:
this is not actually correct if `B` is a different type than `T`

Reply:
I think we should throw an error in that case if it's a differently sized type. Otherwise we would have to do a copy to cast to a larger type.

But I assume that Irrlicht does not do this (and if it did it would be cursed and we want to stop that) because the old code had this optimization as well, so I think it isn't terribly bad to leave this edge case as UB.

Reply:
I agree.
my suggestion:
```cpp
if constexpr (sizeof(T) != sizeof(B)) {
    [[maybe_unused]] auto c_ = static_cast<intptr_t>(c);
    _IRR_DEBUG_BREAK_IF(c_ >= static_cast<intptr_t>(str.data()) && c_ <= static_cast<intptr_t>(str.data() + str.size())); 
}
```

Reply:
sfan5, what the fuck

Reply:
man, i didnt even know constexpr if statements could be used like that, nor did i think maybe_unused could be used on regular values.
My C++ is rusty...

Reply:
How do you want to do a template-level check on a condition that can only be verified at runtime?

Reply:
> How do you want to do a template-level check on a condition that can only be verified at runtime?

You are using constexpr and checking the type at compile-time too, no?

I personally thing this is all being overdone to death. We are likely axing irrString soon anyway so no point in trying to polish the chipped paint

Reply:
@sfan5's code is the way to go here modulo arbitrary style choices, I'm not sure what @swagtoy is on about? `if constexpr` makes perfect sense, the sizes of the types are comptime known after all. `c_` needs to be `[[maybe_unused]]` because the Irrlicht debug macro could be defined to be a no-op. And the check just checks whether the pointer is in range. This check needs to happen at runtime because the pointers obviously won't be known at compile time in general.

Reply:
@sfan5's code doesn't compile.

Reply:
Fixed king